test/server/runner-test.js

```diff
@@ -325,4 +309,140 @@ describe('Runner', function () {
+        });
+
+        it('Should raise an error if browser disconnected during bootstrapping', function (done) {
+            var connection1 = testCafe.createBrowserConnection();
+            var connection2 = testCafe.createBrowserConnection();
+
+            var run = runner
+                .browsers(connection1, connection2)
+                .reporter('list')
+                .src('test/server/data/test-suite/top.test.js')
+                .run()
+                .then(function () {
+                    throw new Error('Promise rejection expected');
+                })
+                .catch(function (err) {
+                    expect(err.message).eql('The Chrome 41.0.2227 / Mac OS X 10.10.1 browser disconnected. ' +
+                                            'This problem may appear when a browser hangs or is closed, ' +
+                                            'or due to network issues.');
+                });
+
+            connection1.HEARTBEAT_TIMEOUT = 200;
+            connection2.HEARTBEAT_TIMEOUT = 200;
+
+            var options = {
+                url:            connection1.url,
+                followRedirect: false,
+                headers:        {
+                    'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 ' +
+                                  '(KHTML, like Gecko) Chrome/41.0.2227.1 Safari/537.36'
+                }
+            };
+
+            request(options);
+
+            run
+                .then(function () {
+                    done();
+                })
+                .catch(done);
+        });
+
+        it('Should raise an error if connection breaks while tests are running', function (done) {
+            var brokenConnection = testCafe.createBrowserConnection();
+
+            brokenConnection.establish('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 ' +
+                                       '(KHTML, like Gecko) Chrome/41.0.2227.1 Safari/537.36');
+
+            var run = runner
+                .browsers(brokenConnection)
+                .reporter('list')
+                .src('test/server/data/test-suite/top.test.js')
+                .run()
+                .then(function () {
+                    throw new Error('Promise rejection expected');
+                })
+                .catch(function (err) {
+                    expect(err.message).eql('I have failed :(');
+                });
+
+            var interval = setInterval(function () {
+                var status = brokenConnection.getStatus();
+
+                if (status.cmd === COMMAND.run) {
+                    clearInterval(interval);
+                    interval = NaN;
+
+                    brokenConnection.emit('error', 'I have failed :(');
+                }
+            }, 200);
+
+            setTimeout( function () {
+                if (!interval.isNaN())
+                    clearInterval(interval);
+
+            }, this.timeout());
```

Comment:
No need for this guard, mocha will automatically fail test if it will not pass withitn 2 sec.